### PR TITLE
Adding --dump-compilation-phases-to= flag and compiler API.

### PR DESCRIPTION
**Pull request by benvanik**
This dumps one file at the end of every compilation phase to a directory. I do this every day by hand/with scripts as a way to speed up triage ("at which phase do we go wrong") and speed up iteration (compile-to a particular phase before I am changing things and then compile-from that point). This should make it easier to ask for people to share IR at various stages.

Example:
`iree-compile --dump-compilation-phases-to=phases/ ...`
image
then you can resume from any phase by name:
`iree-compile --compile-from=global-optimization phases/compiled_unet.4.global-optimization.mlir`

### CODE REVIEW

File: compiler/src/iree/compiler/Pipelines/Pipelines.cpp [hunk 1]
```diff
@@ -129,9 +186,14 @@ void buildIREEPrecompileTransformPassPipeline(
     break;
   default:
     if (compileFrom < IREEVMPipelinePhase::Preprocessing) { // late-entry.
-      // Not a large enough phase for IREE_TRACE_ADD_[BEGIN,END]_FRAME_PASS.
+      IREE_TRACE_ADD_BEGIN_FRAME_PASS(passManager, "Preprocessing");
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Pipelines/Pipelines.cpp:189]: Likely fine to add this back, but the Tracy UI was looking strange with how many frames we defined (lots of rows stacked vertically).

I had this screenshot on https://github.com/openxla/iree/pull/15216:
image

File: compiler/src/iree/compiler/Tools/iree_compile_lib.cc [hunk 2]
```diff
@@ -134,6 +134,11 @@ int mlir::iree_compiler::runIreecMain(int argc, char **argv) {
         compileToPhases.push_back(name.str());
       });
 
+  llvm::cl::opt<std::string> dumpCompilationPhasesTo(
+      "dump-compilation-phases-to",
+      llvm::cl::desc("Dumps IR at the end of each compilation phase to the "
+                     "given directory."));
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Tools/iree_compile_lib.cc:140]: I can follow-up to this with docs here: https://iree.dev/developers/general/developer-tips/#compiling-phase-by-phase

File: compiler/src/iree/compiler/Dialect/Util/Transforms/DumpModule.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  void runOnOperation() override {
+    // Ensure the parent paths exist.
+    llvm::sys::fs::create_directories(llvm::sys::path::parent_path(path));
+
+    // Attempt to open file - should succeed as long as permissions are ok.
+    std::string error;
+    auto file = mlir::openOutputFile(path, &error);
+    if (!file) {
+      llvm::errs() << "while dumping to " << path << ": " << error << "\n";
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/DumpModule.cpp]: nit: quotes around path?
```suggestion
      llvm::errs() << "while dumping to '" << path << "': " << error << "\n";
```

File: compiler/src/iree/compiler/Dialect/Util/Transforms/DumpModule.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      return signalPassFailure();
+    }
+
+    // If going to binary serialize out and otherwise print as text.
+    if (llvm::sys::path::extension(path) == ".mlirbc") {
+      BytecodeWriterConfig config;
+      if (failed(writeBytecodeToFile(getOperation(), file->os(), config))) {
+        llvm::errs() << "failed to serialize module to " << path << "\n";
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/DumpModule.cpp]: if accepting the other suggestion, quotes for consistency? (even though the path can't be empty string here)
```suggestion
        llvm::errs() << "failed to serialize module to '" << path << "'\n";
```

**Review by ScottTodd — APPROVED** [review]

File: compiler/src/iree/compiler/Pipelines/Pipelines.cpp (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Pipelines/Pipelines.cpp:189]: it's just one more line, and if someone is using preprocessing it's useful to see - would be better to fix tracy :P

**Pull request merged**